source/develop/release-management/features/storage/cd-block-storage.md

```diff
@@ -0,0 +1,228 @@
+
+```
+    <ovirt-vm:device devtype="disk" name="sdc"/>
+```
+
+### Bugs
+
+- [Bug 1589763: Error changing CD for a running VM when ISO image is on a block domain](https://bugzilla.redhat.com/1589763)
+- [Bug 1868643: Volume not deactivated after ejecting CDROM](https://bugzilla.redhat.com/1868643)
+
+
+### Functionality added with this feature
+
+This feature adds new APIs to vdsm for loading, ejecting and changing CD.
+With these functions, all aforementioned functionality will work properly.
+
+#### Vdsm
+
+- Add VM.insertCD API
+- Add VM.ejectCD API
+- Updated recovery flow to check `in_progress` flag for CDs
+
+
+### Flows
+
+#### VM with empty tray, insert CD
+
+1. engine sends `insertCD` command, specifying ISO by PDIV format
+1. add new drivespec to VM metadata with new attribute `in_progress`
+```
+    <ovirt-vm:device devtype="disk" name="sdc" in_progress="yes">
+```
+1. prepare new drivespec
+  - if failed:
+    1. remove drivespec (log errors)
+    1. fail the flow with original error
+1. loads CD into VM
+  - if failed:
+    1. teardown the volume (log error)
+    1. remove the drive spec (log error)
+    1. fail the flow with the original error
+1. remove `in-progress` attribute from VM metadata
+  - if failed: log the error and succeed
+
+
+#### VM with loaded CD, eject CD
+
+1. engine sends `ejectCD`, specifying ISO path by PDIV format 
+1. vdsm updates VM metadata with new attribute `in_progress`:
+```
+    <ovirt-vm:device devtype="disk" name="sdc" in_progress="yes">
+```
+1. vdsm eject CD from VM (calling libvirt `updateDeviceFlags()`)
+  - if failed (e.g. timeout?):
+    1. remove the `in_progress` flag (log error)
+    1. fail with original error
+1. tear down drivespec
+  - if tearing down volume fails: log error (volume used by another VM)
+1. remove drivespec from VM metadata
+  - if failed: log error and succeed
+
+
+#### VM recovery flow
+
+Try to cleanup storage leftovers from insert and remove operations without changing VM state.
+
+1. metadata in progress, volume is not active
+  - remove metadata from XML
+1. metadata in progress, volume active, no device
+  - deactivate the volume
+  - remove metadata from XML
+1. metadata in progress, volume active, device exists
+  - remove attribute `in_progress` from metadata
+
+
+#### VM with loaded CD, changeCD
+
+1. engine sends ejectCD, specifying ISO path by PDIV format
+1. engine sends insetrCD command, specifying ISO by PDIV format
+
+##### Changes in Engine
+
+For engine only new internal commands will be added and there will be no changes visible to the user.
+
+- Update ChangeCD to use InsertCD and EjectCD on 4.5 cluster version
+
+On engine side also two new internal commands will be added:
+
+- Add InsertCD sub-command
+- Add EjectCD sub-command
+
+
+## Installation/Upgrade
+
+New vdsm APIs will we delivered with new vdsm packages contaning these changes.
+Changes in engine will require approprate vdsm version and use new vdsm APIs.
```

Comment:
" .. and use .." or " ... to use ..."? Will the new functionality be activated depending on the cluster version?

Reply:
yes, will use new API only for cluster version 4.5 (mentioned in Changes in engine par)